03.bowling/frame.rb

```diff
@@ -0,0 +1,53 @@
+# frozen_string_literal: true
+
+class Frame
+  attr_reader :shots
+
+  def initialize(shots, is_last_frame: false)
+    @shots = shots
+    @is_last_frame = is_last_frame
+  end
+
+  def pins
+    @shots.map(&:pin)
+  end
+
+  def score(index, all_frames)
+    pins.sum + calc_bonus(index, all_frames)
+  end
+
+  def strike?
+    @shots.first.strike?
+  end
+
+  def spare?
+    !strike? && pins[0..1].sum == 10
+  end
+
+  private
+
+  def calc_bonus(index, all_frames)
+    if strike?
+      calc_strike_bonus(index, all_frames)
+    elsif spare?
+      calc_spare_bonus(index, all_frames)
+    else
+      0
+    end
+  end
+
+  def calc_strike_bonus(index, all_frames)
+    if @is_last_frame
+      0
+    else
+      next_shots = all_frames[index + 1].shots + (all_frames[index + 2]&.shots || [])
+      next_shots[0..1].sum(&:pin)
+    end
+  end
+
+  def calc_spare_bonus(index, all_frames)
+    return 0 if @is_last_frame
+
+    all_frames[index + 1].shots.first.pin || 0
```

Comment:
```suggestion
    all_frames[index + 1].shots.first.pin
```

nilの可能性はなくなったはずですー

あと三項演算子でまとめてもいいかもしれません。

```ruby
@is_last_frame ? 0 : all_frames[index + 1].shots.first.pin
```